documentation/src/main/asciidoc/manual-index.asciidoc

```diff
@@ -361,3 +361,281 @@ parameters through - see <<lucene-indexing-performance>> for more details.
+is highly dependent on your overall architecture, database design and even data values.
+
+You should experiment with these settings to find out what's best in your particular case.
+====
+
+===== Threads
+
+The maximum number of threads used by the job execution is defined through method `maxThreads()`.
+Within the N threads given, there’s 1 thread reserved for the core, so only N - 1 threads are
+available for different partitions. If N = 1, the program will work, and all batch elements will run
+in the same thread. The default number of threads used in Hibernate Search is 10. You can overwrite
+it with your preferred number.
+
+====
+[source, JAVA]
+----
+MassIndexingJob.parameters()
+        .maxThreads( 5 )
+        ...
+----
+====
+
+[NOTE]
+====
+Note that the batch runtime cannot guarantee the requested number of threads are available, it will
+use as many as possible up to the requested maximum (JSR352 v1.0 Final Release, page 34). Note also that all
+batch jobs share the same thread pool—so it's a good idea to execute jobs separately.
+====
+
+===== Row per partitions
+
+Each partition consists of a fix number of elements to index. You may tune exactly how many elements
+a partition will hold with `rowsPerPartition`.
+
+[NOTE]
+====
+This property has nothing to do with checkpoint algorithms.
+
+Please see the <<jsr-352-checkpoint,Checkpoint section>> to see how to tune checkpoint algorithms.
+====
+
+When `rowsPerPartition` is low, there will be many small partitions,
+so processing threads will be less likely to starve (stay idle because there's no more partition to process),
+but on the other hand you will only be able to use a small fetch size,
+which will increase the number of database accesses.
+
+When `rowsPerPartition` is high, there will be a few big partitions,
+so you will be able to take advantage of a higher fetch size,
+which will reduce the number of database accesses,
+but on the other hand the processing of each partition will require to hold more objects in memory.
+
+[NOTE]
+====
+Each partition deals with one root entity type, so two different entity types will never run under
+the same partition.
+====
+
+====
+[source, JAVA]
+----
+MassIndexingJob.parameters()
+        .rowsPerPartition( 5000 )
+        ...
+----
+====
+
+==== [[jsr-352-checkpoint]] Checkpoint
+
+The mass indexing job supports checkpoint algorithm, which helps you to restart your suspended or
+failed job from the last successful checkpoint. Assuming that N is the value of checkpoint interval,
+a partition will reach the checkpoint at every N items processed. You can overwrite the default
+interval to adapt your business requirement. By default, it is set to 500.
+
+====
+[source, JAVA]
+----
+MassIndexingJob.parameters()
+        .checkpointInterval( 1000 )
+        ...
+----
+====
+
+Checkpoint algorithms are partition-scoped, meaning each partition has its own checkpoint counter
```

Comment:
What happens when the checkpoint interval is higher than the partition size? Do checkpoints become useless? Maybe we should trigger an exception in this case?